### DISCUSSION TITLE
More information about huidu WF series

### DISCUSSION THREAD

**Discussion by Jimmmy-y** [mrcodetastic/ESP32-HUB75-MatrixPanel-DMA#613] [Ideas]
When I received this board, 
I was surprised to see that it had an ESP32 built in.
I thought it could be used as a development board, 
so I searched for it.

Related URL
https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/issues/433
https://github.com/kreier/huidu_hd-wf4
https://github.com/mrfaptastic/HD-WF1-LED-MatrixPanel-DMA

WF2 and WF4 are driven by ESP32-S3.
(WF1 is ESP32-S2)
This product can run up to 768 x 128
(probably up to 1280 x 256)

This board worked great with the LED module below.
P2
https://ko.aliexpress.com/item/4001321148041.html
P2.5
https://ko.aliexpress.com/item/1005006612553395.html
P2.5 OUTDOOR
https://ko.aliexpress.com/item/1005005020479938.html

Datasheet
WF1
https://huidu-cn.oss-ap-southeast-1.aliyuncs.com/eData/File/Specifications/Single_color/WFX_series/HD-WF1%20Specification%20V0.1.pdf
WF2
https://5.imimg.com/data5/SELLER/Doc/2021/8/HR/TA/BA/134227014/huidu-hd-wf2-led-controller-card.pdf
WF4
https://yuchip-led.com/wp-content/uploads/2021/08/HD-WF4-Specification-V6.0.1.pdf

![image](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/a620abed-4d9e-44cb-8732-f7613b8b67c1)
![KakaoTalk_20240402_192525683](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/5af690df-35aa-4689-a314-a29f13f8064f)
![KakaoTalk_20240402_192108701](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/f950ef16-d342-4ee3-9e16-e9ee068a98ad)
![KakaoTalk_20240402_192108701_01](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/7a6f48be-c1d6-4e54-8efd-e8dface5f17c)
![KakaoTalk_20240402_192108701_02](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/db0d6c20-02cc-45ab-a609-41ad8ec1a540)
![KakaoTalk_20240402_192108701_03](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/8a13a185-40cd-46f2-a6f1-11d128bd050f)
![KakaoTalk_20240402_192108701_04](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/eb48ffc9-b1e8-4914-8346-9423ea08e843)
![KakaoTalk_20240402_192108701_05](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/886f1ae6-4ddc-4689-b36c-aa3569697f16)
![KakaoTalk_20240402_192108701_07](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/5450ea2a-632d-4595-a124-b79fa471ac8f)
![KakaoTalk_20240402_192108701_08](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/6c490200-976b-4559-8513-3d6f77f64122)
![KakaoTalk_20240402_192108701_10](https://github.com/mrfaptastic/ESP32-HUB75-MatrixPanel-DMA/assets/58300422/f5848c3d-ae0c-4940-8bb7-135a50fae219)

**Comment by mrcodetastic** [comment]
Good Intel.

> WF2 and WF4 are driven by ESP32-S3.\n(WF1 is ESP32-S2)\nThis product can run up to 768 x 128\n(probably up to 1280 x 256)

You will not be able to drive such large resolutions using this library unless you reduce the colour depth to like 3bits.

From whatever software that ran on the ESP32-S2 / WF1 version, it only have 4 bit colour and bigbanged the GPIO.

This library uses 24bits per pixel colour depth by default which uses a lot more memory and therefore reduces the effective resolution you can drive with a good (aka. Low visible flicker) refresh rate.

**Reply by mrcodetastic** [reply]
Also be aware then when using this library you can only practically map to one HUB 75 connector. The others could only really be used for general purpose 5V output GPIOs.

Esp32 parallel DMA mode supports max 16 bit (GPIO) parallel output. Only enough parallel bits for one chain of panels via. one connector.

**Reply by board707** [reply]
> Esp32 parallel DMA mode supports max 16 bit (GPIO) parallel output. Only enough parallel bits for one chain of panels via. one connector.
 
Enough for two. Two parallel set of R0G0B0-R1G1B1 (12 pins) and common CLK, LAT and OE.
The A B C D E switches can be easily managed externally.

**Reply by harissutanrafiq** [reply]
> > Esp32 parallel DMA mode supports max 16 bit (GPIO) parallel output. Only enough parallel bits for one chain of panels via. one connector.
> 
> Enough for two. Two parallel set of Regarding (12 pins) and common CLK, LAT and OE. The A B C D E switches can be easily managed externally.

Ahh this what i want, its possible we just need 2  LAT Pin

We just need
R1R2G1G2B1B2
LAT1,LAT2,CLK,OE

when we want display on second connector just use LAT2 PIN.

but i don't know where line in this lib need tobe change. Can you help me ?

**Reply by mrcodetastic** [reply]
So the goal is to drive two parallel sets of panels by alternating the LAT between them?

That's like a bit of a library rewrite...

**Reply by board707** [reply]
> Ahh this what i want, its possible we just need 2 LAT Pin

@harissutanrafiq 
Using a different LAT pins and common color channels (R1R2G1G2B1B2) on two HUBs seems to be completely pointless as for me. As I mentioned in #661, this method doesn't provide any advantages comparing with traditional daisy-chain connection.

> So the goal is to drive two parallel sets of panels by alternating the LAT between them?

@mrcodetastic 
It absolutely useless, I think. The MUST HAVE condition to parallel driving of two panel sets is using a two sets of  color channel (R1R2G1G2B1B2) pins. By using two LAT pins we can only address two sets of panels sequentially.

**Reply by harissutanrafiq** [reply]
> So the goal is to drive two parallel sets of panels by alternating the LAT between them?
> 
> That's like a bit of a library rewrite...

Yes right sir
![IMG_20240729_150721](https://github.com/user-attachments/assets/9d9a3dc6-7720-4985-b8a2-089c50650187)

And still noise

**Reply by harissutanrafiq** [reply]
> > Ahh this what i want, its possible we just need 2 LAT Pin
> 
> @harissutanrafiq Using a different LAT pins and common color channels (R1R2G1G2B1B2) on two HUBs seems to be completely pointless as for me. As I mentioned in #661, this method doesn't provide any advantages comparing with traditional daisy-chain connection.
> 
> > So the goal is to drive two parallel sets of panels by alternating the LAT between them?
> 
> @mrcodetastic It absolutely useless, I think. The MUST HAVE condition to parallel driving of two panel sets is using a two sets of color channel (R1R2G1G2B1B2) pins. By using two LAT pins we can only address two sets of panels sequentially.

But why chinese Huidu controller using two channel ? 

for speed it is not so good, but for effectiveness this pin saving is very useful. I think for 4bit color the difference is not so visible.

**Reply by board707** [reply]
> or speed it is not so good, but for effectiveness this pin saving is very useful

What is the effectiveness in that case?
I don't want to argue with you, connect the panels as you like. I just don't see any point for myself in using this approach.

**Reply by harissutanrafiq** [reply]
> > or speed it is not so good, but for effectiveness this pin saving is very useful
> 
> What is the effectiveness in that case? I don't want to argue with you, connect the panels as you like. I just don't see any point for myself in using this approach.

Ah ok sir, any way thanks for your fast response. And i love your lib using dmd rgb stm32 👌

**Reply by harissutanrafiq** [reply]
> > or speed it is not so good, but for effectiveness this pin saving is very useful
> 
> What is the effectiveness in that case? I don't want to argue with you, connect the panels as you like. I just don't see any point for myself in using this approach.

Just update my project, it works now

![IMG_20240806_072902](https://github.com/user-attachments/assets/4c01e9d4-6143-4693-b9df-9510cc82da2d)

![IMG_20240806_074549](https://github.com/user-attachments/assets/d7f3b8e0-1e6f-4813-befc-b834e0f5898a)
